cdc/sink/producer/kafka/config.go

```diff
@@ -0,0 +1,297 @@
+	// partitions, this could prevent potential correctness problems.
+	if c.PartitionNum > realPartitionCount {
+		return cerror.ErrKafkaInvalidPartitionNum.GenWithStack(
+			"the number of partition (%d) specified in sink-uri is more than that of actual topic (%d)",
+			c.PartitionNum, realPartitionCount)
+	}
+	return nil
+}
+
+// newSaramaConfig return the default config and set the according version and metrics
+func newSaramaConfig(ctx context.Context, c *Config) (*sarama.Config, error) {
+	config := sarama.NewConfig()
+
+	version, err := sarama.ParseKafkaVersion(c.Version)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaInvalidVersion, err)
+	}
+	var role string
+	if util.IsOwnerFromCtx(ctx) {
+		role = "owner"
+	} else {
+		role = "processor"
+	}
+	captureAddr := util.CaptureAddrFromCtx(ctx)
+	changefeedID := util.ChangefeedIDFromCtx(ctx)
+
+	config.ClientID, err = kafkaClientID(role, captureAddr, changefeedID, c.ClientID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	config.Version = version
+	// See: https://kafka.apache.org/documentation/#replication
+	// When one of the brokers in a Kafka cluster is down, the partition leaders
+	// in this broker is broken, Kafka will election a new partition leader and
+	// replication logs, this process will last from a few seconds to a few minutes.
+	// Kafka cluster will not provide a writing service in this process.
+	// Time out in one minute.
+	config.Metadata.Retry.Max = 120
+	config.Metadata.Retry.Backoff = 500 * time.Millisecond
+	// If it is not set, this means a metadata request against an unreachable
+	// cluster (all brokers are unreachable or unresponsive) can take up to
+	// `Net.[Dial|Read]Timeout * BrokerCount * (Metadata.Retry.Max + 1) +
+	// Metadata.Retry.Backoff * Metadata.Retry.Max`
+	// to fail.
+	// See: https://github.com/Shopify/sarama/issues/765
+	// and https://github.com/pingcap/ticdc/issues/3352.
+	config.Metadata.Timeout = 1 * time.Minute
+
+	config.Producer.Partitioner = sarama.NewManualPartitioner
+	config.Producer.MaxMessageBytes = c.MaxMessageBytes
+	config.Producer.Return.Successes = true
+	config.Producer.Return.Errors = true
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	// Time out in five minutes(600 * 500ms).
+	config.Producer.Retry.Max = 600
+	config.Producer.Retry.Backoff = 500 * time.Millisecond
+	switch strings.ToLower(strings.TrimSpace(c.Compression)) {
+	case "none":
+		config.Producer.Compression = sarama.CompressionNone
+	case "gzip":
+		config.Producer.Compression = sarama.CompressionGZIP
+	case "snappy":
+		config.Producer.Compression = sarama.CompressionSnappy
+	case "lz4":
+		config.Producer.Compression = sarama.CompressionLZ4
+	case "zstd":
+		config.Producer.Compression = sarama.CompressionZSTD
+	default:
+		log.Warn("Unsupported compression algorithm", zap.String("compression", c.Compression))
+		config.Producer.Compression = sarama.CompressionNone
+	}
+
+	// Time out in one minute(120 * 500ms).
```

Comment:
Curious about, if each retry reaches the timeout (20s defined in L273), will the total retry time be 120 * 20s?

Reply:
It should be a total of 20s.

Reply:
I'll reset it.